### PR TITLE
OWGrep: Add widget

### PR DESCRIPTION
**Pull request by janezd**
A widget for getting the data from a text file, like a log.

##### Includes
- [X] Code changes
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: orangecontrib/prototypes/widgets/owgrep.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        data = Output("Data", Table, default=True)
+
+    text = Setting("")
+    skip_lines = Setting(0)
+    use_lines = Setting(0)
+    column_names = Setting(True)
+    auto_send = Setting(True)
+    find_text = Setting("")
+
+    re_number = re.compile(r"^([+-]?\d+([.,]\d+)?)$")
```
Comment (kernc) [inline orangecontrib/prototypes/widgets/owgrep.py]: doesn't match -.3

Comment (kernc) [reply orangecontrib/prototypes/widgets/owgrep.py]: Outer parentheses are unnecessary.

File: orangecontrib/prototypes/widgets/owgrep.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        out_view.setMinimumSize(500, 250)
+        out_view.setLineWrapMode(out_view.NoWrap)
+        out_view.setReadOnly(True)
+        out_view.setFontFamily("Courier")
+        self.mainArea.layout().addWidget(out_view)
+
+        self.file_changed()
+
+    def file_changed(self):
+        name = "/Users/janez/Desktop/out.out"
```
Comment (ajdapretnar) [inline orangecontrib/prototypes/widgets/owgrep.py]: For testing, this should be disabled.

File: orangecontrib/prototypes/widgets/owgrep.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        for line in lines:
+            if self.text in line:
+                for i in range(self.skip_lines):
+                    line = next(lines)
+                if self.column_names:
+                    if column_names is None:
+                        column_names = line.split()
+                    line = next(lines)
+                for i in range(self.use_lines):
+                    data.append(line.split())
```
Comment (kernc) [inline orangecontrib/prototypes/widgets/owgrep.py]: Love the undocumented state machine!

So this only splits matching lines by whitespace?

What if different lines have different number of tokens?

File: orangecontrib/prototypes/widgets/owgrep.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    def text_changed(self):
+        self.read_file()
+
+    def read_file(self):
+        data = []
+        lines = iter(self.file_lines)
+        column_names = None
+        for line in lines:
+            if self.text in line:
```
Comment (kernc) [inline orangecontrib/prototypes/widgets/owgrep.py]: Why can't `text` be a regex?

File: orangecontrib/prototypes/widgets/owgrep.py (hunk 2 above)
Comment (jerneju) [reply orangecontrib/prototypes/widgets/owgrep.py]: `FileNotFoundError` and I cannot open the widget.

File: orangecontrib/prototypes/widgets/owgrep.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+from AnyQt.QtCore import Qt
+from AnyQt.QtGui import QKeySequence
+from AnyQt.QtWidgets import QTextEdit, QAction
+
+from Orange.data import (Table, Domain, ContinuousVariable, DiscreteVariable,
+                         StringVariable)
+from Orange.widgets import gui
+from Orange.widgets.settings import Setting
+from Orange.widgets.widget import OWWidget, Msg, Output
+
```
Comment (jerneju) [inline orangecontrib/prototypes/widgets/owgrep.py]: Many other widgets have two newlines here.

File: orangecontrib/prototypes/widgets/owgrep.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        in_view.setLineWrapMode(in_view.NoWrap)
+        in_view.setReadOnly(True)
+        in_view.setFontFamily("Courier")
+        self.mainArea.layout().addWidget(in_view)
+
+        copy = QAction("Copy", in_view, triggered=lambda *x: print("foo") or in_view.copy())
+        copy.setShortcuts(QKeySequence(Qt.Key_C | Qt.ControlModifier))
+        in_view.addAction(copy)
+        for action in in_view.actions():
+            print(action.text())
```
Comment (jerneju) [inline orangecontrib/prototypes/widgets/owgrep.py]: This is probably here for debugging purposes.

**Comment by janezd** [comment]
Oh, oh, stop it guys. :) That's something I quickly put together to show Feri how to make new widgets. It's a long long way from being done. I've just put it here to share it with Feri.

Big apologies for forgetting adding a WIP tag. No need to review it at this phase.

**Comment by ajdapretnar** [comment]
I don't think there has been a widget in the history of Orange that was reviewed as quickly as this one. 😆 We jumped at it like hungry wolves. 🐺

**Comment by jerneju** [comment]
I cannot open file.

```python
Traceback (most recent call last):
  File "/prototypes/widgets/owgrep.py", line 189, in browse_file
    self.open_file()
  File "/prototypes/widgets/owgrep.py", line 208, in open_file
    text = f.read()
  File "/python3.5/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```

+1 for the tests

**Pull request merged**